lib/ruby_indexer/lib/ruby_indexer/prefix_tree.rb

```diff
@@ -0,0 +1,103 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyIndexer
+  class PrefixTree
+    extend T::Sig
+
+    sig { params(items: T::Array[String]).void }
+    def initialize(items)
+      @root = T.let(Node.new("", ""), Node)
+
+      items.each { |item| insert(item) }
+    end
+
+    sig { params(prefix: String).returns(T::Array[String]) }
+    def search(prefix)
+      node = T.let(@root, Node)
+
+      prefix.each_char do |char|
+        snode = node.children[char]
+        return [] unless snode
+
+        node = snode
+      end
+
+      node.collect
+    end
+
+    sig { params(item: String).void }
+    def insert(item)
+      node = T.let(@root, Node)
+
+      item.each_char do |char|
+        node = node.children[char] ||= Node.new(node.value + char, char, node)
+      end
+
+      node.leaf = true
+    end
+
+    sig { params(item: String).void }
+    def delete(item)
+      node = T.let(@root, Node)
+
+      # Find the node to be deleted
+      item.each_char do |char|
```

Comment:
I think we can extract this as a private `get_node` method and share it between the `delete` and `search`